pkg/ruler/manager_test.go

```diff
@@ -7,143 +7,111 @@ package ruler
+	t.Run("calling SyncRuleGroups() with an empty map stops all managers", func(t *testing.T) {
+		m.SyncRuleGroups(ctx, nil)
 
-	// Passing empty map / nil stops all managers.
-	m.SyncRuleGroups(context.Background(), nil)
-	require.Nil(t, getManager(m, user1))
+		// Ensure the ruler manager has been stopped for all users.
+		assertManagerMockStopped(t, initialUser1Manager)
+		assertManagerMockStopped(t, initialUser2Manager)
+		assertManagerMockNotRunningForUser(t, m, user1)
+		assertManagerMockNotRunningForUser(t, m, user2)
 
-	// Make sure old manager was stopped.
-	test.Poll(t, 1*time.Second, false, func() interface{} {
-		return mgr1.(*mockRulesManager).running.Load()
-	})
+		// Ensure the files have been removed from disk.
+		assertRuleGroupsMappedOnDisk(t, m, user1, nil)
+		assertRuleGroupsMappedOnDisk(t, m, user2, nil)
 
-	test.Poll(t, 1*time.Second, false, func() interface{} {
-		return mgr2.(*mockRulesManager).running.Load()
+		// Check metrics.
+		assert.Equal(t, 0.0, promtest.ToFloat64(m.managersTotal))
 	})
 
-	// Verify that local rule groups were removed.
-	{
-		users, err := m.mapper.users()
-		require.NoError(t, err)
-		require.Equal(t, []string(nil), users)
-	}
+	t.Run("calling SyncRuleGroups() with the previous config restores the managers", func(t *testing.T) {
+		m.SyncRuleGroups(ctx, map[string]rulespb.RuleGroupList{
+			user1: {user1Group1},
+			user2: {user2Group1},
+		})
 
-	// Resync same rules as before. Previously this didn't restart the manager.
-	m.SyncRuleGroups(context.Background(), userRules)
+		// Ensure the ruler manager has been started.
+		currUser1Manager := assertManagerMockRunningForUser(t, m, user1)
+		currUser2Manager := assertManagerMockRunningForUser(t, m, user2)
+		assert.NotEqual(t, currUser1Manager, initialUser1Manager)
+		assert.NotEqual(t, currUser2Manager, initialUser2Manager)
 
-	newMgr := getManager(m, user1)
-	require.NotNil(t, newMgr)
-	require.True(t, mgr1 != newMgr)
+		// Ensure the files have been mapped to disk.
+		assertRuleGroupsMappedOnDisk(t, m, user1, rulespb.RuleGroupList{user1Group1})
+		assertRuleGroupsMappedOnDisk(t, m, user2, rulespb.RuleGroupList{user2Group1})
 
-	test.Poll(t, 1*time.Second, true, func() interface{} {
-		return newMgr.(*mockRulesManager).running.Load()
+		// Check metrics.
+		assert.Equal(t, 2.0, promtest.ToFloat64(m.managersTotal))
 	})
 
-	// Verify that user rule groups are cached locally again.
-	{
-		users, err := m.mapper.users()
-		require.NoError(t, err)
-		require.Contains(t, users, user1)
-	}
+	t.Run("calling Stop() should stop all managers", func(t *testing.T) {
+		// Pre-condition check.
+		currUser1Manager := assertManagerMockRunningForUser(t, m, user1)
+		currUser2Manager := assertManagerMockRunningForUser(t, m, user2)
 
-	m.Stop()
+		m.Stop()
 
-	test.Poll(t, 1*time.Second, false, func() interface{} {
-		return newMgr.(*mockRulesManager).running.Load()
+		assertManagerMockStopped(t, currUser1Manager)
+		assertManagerMockStopped(t, currUser2Manager)
+
```

Comment:
Nit: shall we add this?

```suggestion

		assertManagerMockNotRunningForUser(t, m, user1)
		assertManagerMockNotRunningForUser(t, m, user2)
```

Reply:
Yes. This made me realised we were not cleaning up `r.userManagers`, which I've done in: https://github.com/grafana/mimir/pull/4994/commits/ab61427d66684d841048d409baf446130b0eed31

Reply:
👍 Nice.